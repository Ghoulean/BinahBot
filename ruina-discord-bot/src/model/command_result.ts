import { DiscordEmbed } from "./discord/discord_embed";

// TODO: The term "command" is overloaded to refer to a registered Discord
// slash command, and to the execution of such command. Note that "autocomplete"
// is a subset of the first, but not of the second.
export interface BaseCommandResult {
    success: boolean;
    error?: string;
};

export interface CommandResult extends BaseCommandResult {
    embed?: DiscordEmbed
}

export interface AutocompleteResult extends BaseCommandResult {
    choices?: string[]
}