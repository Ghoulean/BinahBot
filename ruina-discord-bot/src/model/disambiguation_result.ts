import { Localization, LookupResult } from "@ghoulean/ruina-common";

// TODO: This model is duplicated between the reparser and this package.
// Reconsider where it should be placed
export type DisambiguationResults = {
    id: string;
    locale: Localization;
    query: string;
    lookupResults: LookupResult[];
};