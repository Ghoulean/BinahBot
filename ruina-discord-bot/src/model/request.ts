import { Chapter, Localization } from "@ghoulean/ruina-common";

export type Request = {
    command: string;
    commandArgs: RequestCommandArgs;
    interactionToken: string;
    locale: Localization;
    chapter: Chapter;
    autocomplete?: boolean;
};

export type RequestCommandArgs = {
    [key: string]: string | number | boolean
};
