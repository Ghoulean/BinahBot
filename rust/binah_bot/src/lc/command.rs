use std::str::FromStr;

use lobocorp::lobocorp_common::game_objects::abnormality::EncyclopediaInfo;
use lobocorp::lobocorp_common::localizations::common::Locale;
use lobocorp::lobocorp_reparser::get_encyclopedia_info;

use crate::lc::button::build_buttons;
use crate::lc::button::Code;
use crate::lc::transformers::transform_donttouchme;
use crate::lc::transformers::transform_normal_info;
use crate::lc::transformers::transform_tool_info;
use crate::macros::cast_enum_variant;
use crate::models::binahbot::BinahBotEnvironment;
use crate::models::binahbot::BinahBotLocale;
use crate::models::discord::AllowedMentions;
use crate::models::discord::DiscordEmbed;
use crate::models::discord::DiscordInteraction;
use crate::models::discord::DiscordInteractionData;
use crate::models::discord::DiscordInteractionOptionValue;
use crate::models::discord::DiscordInteractionResponseMessage;
use crate::models::discord::DiscordInteractionResponseType;
use crate::models::discord::DiscordMessageFlag;
use crate::models::discord::MessageResponse;
use crate::utils::get_binahbot_locale;
use crate::utils::get_option_value;

pub fn lc_command(interaction: &DiscordInteraction, env: &BinahBotEnvironment) -> MessageResponse {
    let command_args = interaction
        .data
        .as_ref()
        .and_then(|x| cast_enum_variant!(x, DiscordInteractionData::ApplicationCommand))
        .and_then(|x| x.options.as_ref())
        .unwrap();

    tracing::info!("Lc command: command args: {:#?}", command_args);

    let binah_locale: BinahBotLocale = get_binahbot_locale(interaction);

    let locale: Locale = get_option_value("locale", command_args)
        .and_then(|x| cast_enum_variant!(x, DiscordInteractionOptionValue::String))
        .and_then(|x| Locale::from_str(x.as_str()).ok())
        .unwrap_or(Locale::from(binah_locale.clone()));

    let query = get_option_value("query", command_args)
        .and_then(|x| cast_enum_variant!(x, DiscordInteractionOptionValue::String))
        .and_then(|x| {
            x.parse::<u32>()
                .ok()
                .or_else(|| lobocorp::lobocorp_index::query(x).first().copied())
        });
    let query = match query {
        Some(x) => x,
        None => todo!(),
    };

    let entry = get_encyclopedia_info(&query).expect("couldn't find entry");

    let embed: DiscordEmbed = match &entry {
        EncyclopediaInfo::Normal(x) => transform_normal_info(&x, &locale, &binah_locale, env),
        EncyclopediaInfo::Tool(x) => transform_tool_info(&x, &locale, &binah_locale, env),
        EncyclopediaInfo::DontTouchMe(x) => transform_donttouchme(&x, &locale, &binah_locale, env),
    };

    let is_private = get_option_value("private", command_args)
        .and_then(|x| cast_enum_variant!(x, DiscordInteractionOptionValue::Bool))
        .is_some_and(|x| *x);

    let flags = is_private.then_some(DiscordMessageFlag::EphemeralMessage as i32);

    let components = build_buttons(query, &locale, &binah_locale, &(Code::Encyclopedia, 0), env);

    MessageResponse {
        r#type: DiscordInteractionResponseType::ChannelMessageWithSource,
        data: Some(DiscordInteractionResponseMessage {
            allowed_mentions: Some(AllowedMentions { parse: Vec::new() }),
            content: None,
            embeds: Some(vec![embed]),
            flags: flags,
            components: Some(components),
        }),
    }
}

#[cfg(test)]
mod tests {
    use lobocorp::lobocorp_reparser::get_abno_localization;
    use lobocorp::lobocorp_reparser::get_all_encyclopedia_ids;
    use strum::IntoEnumIterator;

    use crate::models::discord::DiscordApplicationCommandInteractionData;
    use crate::models::discord::DiscordInteractionOptions;
    use crate::models::discord::DiscordInteractionType;
    use crate::models::discord::DiscordUser;
    use crate::test_utils::build_mocked_binahbot_env;

    use super::*;

    #[test]
    fn check_no_crashing() {
        let env = &build_mocked_binahbot_env();
        Locale::iter().for_each(|locale| {
            get_all_encyclopedia_ids()
                .iter()
                .flat_map(|x| get_encyclopedia_info(x))
                .map(|x| {
                    let id = match x {
                        EncyclopediaInfo::Normal(x) => &x.id,
                        EncyclopediaInfo::Tool(x) => &x.id,
                        EncyclopediaInfo::DontTouchMe(x) => &x.id,
                    };
                    get_abno_localization(id, &locale)
                        .map(|x| x.name)
                        .expect("no name")
                })
                .map(|name| build_discord_interaction(name.to_string(), locale.clone()))
                .for_each(|interaction| {
                    let _does_not_crash = lc_command(&interaction, env);
                })
        });
    }

    fn build_discord_interaction(query_string: String, locale: Locale) -> DiscordInteraction {
        DiscordInteraction {
            id: "id".to_string(),
            application_id: "app_id".to_string(),
            r#type: DiscordInteractionType::ApplicationCommand,
            data: Some(DiscordInteractionData::ApplicationCommand(
                DiscordApplicationCommandInteractionData {
                    id: "id".to_string(),
                    name: "lc".to_string(),
                    options: Some(vec![
                        DiscordInteractionOptions {
                            name: "query".to_string(),
                            name_localizations: None,
                            value: DiscordInteractionOptionValue::String(query_string),
                            focused: None,
                        },
                        DiscordInteractionOptions {
                            name: "locale".to_string(),
                            name_localizations: None,
                            value: DiscordInteractionOptionValue::String(locale.to_string()),
                            focused: None,
                        },
                    ]),
                },
            )),
            channel_id: None,
            token: "token".to_string(),
            locale: None,
            guild_locale: None,
            user: Some(DiscordUser {
                id: "snowflake".to_string(),
                username: "username".to_string(),
                avatar: Some("hash".to_string()),
            }),
            member: None,
            message: None,
        }
    }
}
