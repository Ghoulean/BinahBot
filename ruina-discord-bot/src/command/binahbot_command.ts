import { CommandResult } from "../model/command_result";
import { DiscordEmbed } from "../model/discord/discord_embed";
import { Request } from "../model/request";

// Very slightly tinted purple
const BINAH_EMBED_COLOR: number = 0x12001f;

const GITHUB_REPO = "https://github.com/Ghoulean/Ruina/tree/main";
const INVITE_LINK_INDEX = 1;

const ABOUT_BINAHBOT_BASE_EMBED: DiscordEmbed = {
    color: BINAH_EMBED_COLOR,
    title: "About BinahBot:",
    fields: [
        {
            name: "Github repo",
            value: `[Link](${GITHUB_REPO})`,
            inline: true,
        },
        {
            name: "Invite BinahBot to your server:",
            value: "Replace this value",
            inline: true,
        },
    ],
};

// TODO: consider command interface
export class BinahBotCommand {
    private readonly discordEmbed: DiscordEmbed;

    constructor(clientId: string) {
        this.discordEmbed = this.constructDiscordEmbed(clientId);
    }

    public invoke(_request: Request): CommandResult {
        return {
            success: true,
            payload: this.discordEmbed,
        };
    }

    private constructDiscordEmbed(clientId: string): DiscordEmbed {
        const inviteLink: string = `https://discord.com/api/oauth2/authorize?client_id=${clientId}&permissions=0&scope=applications.commands%20bot`;
        const displayLink: string = `[Link](${inviteLink})`
        const retVal: DiscordEmbed = {
            ...ABOUT_BINAHBOT_BASE_EMBED,
        };
        retVal.fields[INVITE_LINK_INDEX].value = displayLink;
        return retVal;
    }
}
