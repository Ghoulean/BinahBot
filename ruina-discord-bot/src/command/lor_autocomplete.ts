import { LookupResult } from "@ghoulean/ruina-common";
import { DataAccessor } from "../accessor/data_accessor";
import { CommandResult } from "../model/command_result";
import { Request } from "../model/request";

// TODO: think about where these constants go
const QUERY_COMMAND_ARG = "query";
const AUTOCOMPLETE_LIMIT = 5;

// TODO: consider interfacing autocomplete function
export class LorAutocomplete {
    private readonly dataAccessor: DataAccessor;

    constructor(dataAccessor: DataAccessor) {
        this.dataAccessor = dataAccessor;
    }

    public autocomplete(request: Request): CommandResult {
        const query: string = request.commandArgs[QUERY_COMMAND_ARG] as string;
        const autocompleteResults: string[] = this.dataAccessor
            .autocomplete(query)
            .slice(0, AUTOCOMPLETE_LIMIT)
            .map((str: string) => {
                return this.dataAccessor.lookup(str, request.locale);
            })
            .map((lookupResult: LookupResult) => {
                return lookupResult.displayQuery ?? lookupResult.query;
            });

        return {
            success: true,
            payload: autocompleteResults,
        };
    }
}
