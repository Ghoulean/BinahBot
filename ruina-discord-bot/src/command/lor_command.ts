import {
    DecoratedAbnoPage,
    DecoratedCombatPage,
    DecoratedKeyPage,
    DecoratedPassive,
    Localization,
    LookupResult,
    PageType,
} from "@ghoulean/ruina-common";
import { DataAccessor } from "../accessor/data_accessor";
import { CommandResult } from "../model/command_result";
import { DisambiguationResults } from "../model/disambiguation_result";
import { DiscordEmbed } from "../model/discord/discord_embed";
import { Request } from "../model/request";
import { EmbedTransformer } from "../transformers/embed_transformer";

const QUERY_COMMAND_ARG = "query";
const LOCALE_COMMAND_ARG = "locale";

// TODO: consider interfacing invoke function
export class LorCommand {
    private readonly dataAccessor: DataAccessor;
    private readonly embedTransformer: EmbedTransformer;

    constructor(
        dataAccessor: DataAccessor,
        embedTransformer: EmbedTransformer
    ) {
        this.dataAccessor = dataAccessor;
        this.embedTransformer = embedTransformer;
    }

    public invoke(request: Request): CommandResult {
        const query: string = request.commandArgs[QUERY_COMMAND_ARG] as string;
        const cardLocale: Localization =
            (request.commandArgs[LOCALE_COMMAND_ARG] as Localization) ??
            request.locale;

        console.log(
            `LorCommand: Received query ${query} with cardLocale ${cardLocale.toString()}`
        );

        let lookupResult: LookupResult;
        try {
            lookupResult = this.dataAccessor.lookup(query, cardLocale);
        } catch (e: unknown) {
            return {
                success: true,
                embed: this.embedTransformer.noResultsFoundEmbed(
                    query,
                    request.locale
                ),
            };
        }

        console.log(
            `LorCommand: Received LookupResult=${JSON.stringify(lookupResult)}`
        );

        const pageType: PageType = lookupResult.pageType;
        let embed: DiscordEmbed;
        switch (pageType) {
            case PageType.ABNO_PAGE: {
                let decoratedAbnoPage: DecoratedAbnoPage;
                try {
                    decoratedAbnoPage = this.dataAccessor.getDecoratedAbnoPage(
                        lookupResult.pageId,
                        cardLocale
                    );
                } catch (e: unknown) {
                    return {
                        success: false,
                        error: `Error occurred while calling getDecoratedAbnoPage: ${JSON.stringify(
                            e
                        )}`,
                    };
                }
                embed = this.embedTransformer.transformAbnoPage(
                    decoratedAbnoPage,
                    request.locale
                );
                break;
            }
            case PageType.COMBAT_PAGE: {
                let decoratedCombatPage: DecoratedCombatPage;
                try {
                    decoratedCombatPage =
                        this.dataAccessor.getDecoratedCombatPage(
                            lookupResult.pageId,
                            cardLocale
                        );
                } catch (e: unknown) {
                    return {
                        success: false,
                        error: `Error occurred while calling getDecoratedCombatPage: ${JSON.stringify(
                            e
                        )}`,
                    };
                }
                embed = this.embedTransformer.transformCombatPage(
                    decoratedCombatPage,
                    request.locale
                );
                break;
            }
            case PageType.KEY_PAGE: {
                let decoratedKeyPage: DecoratedKeyPage;
                try {
                    decoratedKeyPage = this.dataAccessor.getDecoratedKeyPage(
                        lookupResult.pageId,
                        cardLocale
                    );
                } catch (e: unknown) {
                    return {
                        success: false,
                        error: `Error occurred while calling getDecoratedKeyPage: ${JSON.stringify(
                            e
                        )}`,
                    };
                }
                embed = this.embedTransformer.transformKeyPage(
                    decoratedKeyPage,
                    request.locale
                );
                break;
            }
            case PageType.PASSIVE: {
                let decoratedPassive: DecoratedPassive;
                try {
                    decoratedPassive = this.dataAccessor.getDecoratedPassive(
                        lookupResult.pageId,
                        cardLocale
                    );
                } catch (e: unknown) {
                    return {
                        success: false,
                        error: `Error occurred while calling getDecoratedPassive: ${JSON.stringify(
                            e
                        )}`,
                    };
                }
                embed = this.embedTransformer.transformPassive(
                    decoratedPassive,
                    request.locale
                );
                break;
            }
            case PageType.DISAMBIGUATION: {
                let disambiguation: DisambiguationResults;
                try {
                    disambiguation = this.dataAccessor.getDisambiguationResult(
                        lookupResult.pageId
                    );
                } catch (e: unknown) {
                    return {
                        success: false,
                        error: `Error occurred while calling getDisambiguationResult: ${JSON.stringify(
                            e
                        )}`,
                    };
                }
                embed = this.embedTransformer.transformDisambiguationPage(
                    disambiguation,
                    request.locale
                );
                break;
            }
        }

        return {
            success: true,
            embed: embed,
        };
    }
}
