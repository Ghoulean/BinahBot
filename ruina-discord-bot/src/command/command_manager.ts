import { CommandResult } from "../model/command_result";
import { Request } from "../model/request";
import { LorCommand } from "./lor_command";

export class CommandManager {
    private readonly lorCommand: LorCommand;

    constructor(lorCommand: LorCommand) {
        this.lorCommand = lorCommand;
    }

    public async routeRequest(request: Request): Promise<CommandResult> {
        switch (request.command) {
            // TODO: consider moving to enum
            case "lor":
                console.log(`CommandManager: Routing request to LOR`);
                return await this.lorCommand.invoke(request);
            default:
                console.log(`CommandManager: Did not find route for ${request.command}`);
                return {
                    success: false,
                    error: `Command for ${request.command} not found`,
                };
        }
    }
}
