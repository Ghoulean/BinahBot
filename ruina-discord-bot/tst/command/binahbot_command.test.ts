import { Chapter, Localization } from "@ghoulean/ruina-common";
import { BinahBotCommand } from "../../src/command/binahbot_command";
import { Request } from "../../src/model/request";

const REQUEST: Request = {
    command: "binahbot",
    commandArgs: {},
    interactionToken: "it",
    locale: Localization.ENGLISH,
    chapter: Chapter.IMPURITAS_CIVITATIS,
};

const CLIENT_ID: string = "clientId";

let binahbotCommand: BinahBotCommand;

beforeEach(() => {
    binahbotCommand = new BinahBotCommand(CLIENT_ID);
});

test("should return about me", () => {
    expect(binahbotCommand.invoke(REQUEST)).toMatchSnapshot();
});
