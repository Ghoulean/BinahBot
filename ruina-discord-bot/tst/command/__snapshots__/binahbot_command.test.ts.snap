// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return about me 1`] = `
{
  "embed": {
    "color": 1179679,
    "fields": [
      {
        "inline": true,
        "name": "Github repo",
        "value": "[Link](https://github.com/Ghoulean/Ruina/tree/main)",
      },
      {
        "inline": true,
        "name": "Invite BinahBot to your server:",
        "value": "[Link](https://discord.com/api/oauth2/authorize?client_id=clientId&permissions=0&scope=applications.commands%20bot)",
      },
    ],
    "title": "About BinahBot:",
  },
  "success": true,
}
`;
